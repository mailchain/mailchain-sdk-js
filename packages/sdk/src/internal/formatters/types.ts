export type MailAddress = {
	name: string;
	address: string;
};

export type MailData = {
	id: string;
	subject: string;
	from: MailAddress;
	date: Date;
	recipients: MailAddress[];
	carbonCopyRecipients: MailAddress[];
	blindCarbonCopyRecipients: MailAddress[];
	message: string;
	plainTextMessage: string;
};
