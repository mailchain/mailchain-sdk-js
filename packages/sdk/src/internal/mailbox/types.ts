import { PublicKey } from '@mailchain/crypto';

export type MessagePreview = {
	mailbox: PublicKey;
	messageId: string;
	from: string;
	owner: string;
	subject: string;
	snippet: string;
	isRead: boolean;
	systemLabels: SystemMessageLabel[];
	hasAttachment: boolean;
	timestamp: Date;
	to: string[];
	cc: string[];
	bcc: string[];
};

export type Message = {
	from: string;
	replyTo?: string;
	to: string[];
	cc: string[];
	bcc: string[];
	subject: string;
	timestamp: Date;
	body: string;
};

/** Copy from cmd/inbox/internal/datastore/labels.go */
export const SYSTEM_MESSAGE_LABELS = [
	'archive',
	'draft',
	'inbox',
	'sent',
	'spam',
	'starred',
	'trash',
	'unread',
	'outbox',
] as const;
export type SystemMessageLabel = (typeof SYSTEM_MESSAGE_LABELS)[number];
export type UserMessageLabel = string;
