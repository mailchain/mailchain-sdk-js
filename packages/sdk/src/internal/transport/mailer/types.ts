import { MailAddress } from '../mail/types';

export type MailerData = {
	/**
	 * Message subject line, this will become the `Subject:` header in the email.
	 */
	subject: string;
	/**
	 * Reply-To address, this will become the `Reply-To:` header in the email.
	 */
	replyTo?: MailAddress;
	/**
	 * HTML formatted message, this will become the `text/html` part of the email.
	 */
	message: string;
	/**
	 * Plain text formatted message, this will become the `text/plain` part of the email.
	 */
	plainTextMessage: string;
};
