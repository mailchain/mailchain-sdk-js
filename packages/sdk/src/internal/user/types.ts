import { MailchainAddress, ProtocolType } from '@mailchain/addressing';
import { PublicKey } from '@mailchain/crypto';

/**
 * Represents different type of alias to be set for the {@link UserMailbox}.
 */
export type Alias = {
	address: MailchainAddress;
	isDefault: boolean;
	allowSending: boolean;
	allowReceiving: boolean;
};

/**
 * Entity for sending and receiving messages based on the provided identity key and messaging key.
 */
export type UserMailbox = {
	id: string;
	/**
	 * - `'account'` - Mailchain account (example: `'alice@mailchain.com'`).
	 * - `'wallet'` - registered mailbox based blockchain wallet identity key (example: `'0xEBaae0532dF65ee3f1623f324C9620bB84c8af8d@ethereum.mailchain.com'`)
	 */
	type: 'account' | 'wallet';
	identityKey: PublicKey;
	/** The user preferred label to shown for this mailbox. If `null`, no user preferred label is defined, the application is free to compute one at runtime. */
	label: string | null;
	/** Will contain at least one defined {@link Alias}. */
	aliases: [Alias, ...Alias[]];
	/** Ingredients for the creation of messaging private key via the keyring. */
	messagingKeyParams: {
		address: Uint8Array;
		protocol: ProtocolType;
		network: string;
		nonce: number;
	};
};
