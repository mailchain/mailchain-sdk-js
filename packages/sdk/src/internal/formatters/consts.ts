export const X_IDENTITY_KEYS = 'X-IdentityKeys' as const;
