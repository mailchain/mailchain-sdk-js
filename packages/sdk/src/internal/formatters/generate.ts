import * as mimetext from 'mimetext';
import { MailAddress, MailData } from './types';

type EncodedMailData = string;

export const createMimeMessage = (
	mailData: MailData,
): {
	original: EncodedMailData;
	visibleRecipients: EncodedMailData;
	blindRecipients: { recipient: MailAddress; content: EncodedMailData }[];
} => {
	const msg = mimetext.createMimeMessage();
	msg.setHeader('Message-ID', mailData.id);
	msg.setHeader('Date', mailData.date.toISOString());
	msg.setSender({ name: mailData.from.name, addr: mailData.from.address });
	msg.setRecipient(mailData.recipients.map((rec) => ({ name: rec.name, addr: rec.address })));
	// @ts-ignore - for some reason ts complains about setCc
	msg.setCc(mailData.carbonCopyRecipients.map((rec) => ({ name: rec.name, addr: rec.address })));
	msg.setSubject(encodeURIComponent(mailData.subject));
	msg.setMessage('text/html', encodeURIComponent(mailData.message));

	const visibleRecipients = msg.asRaw();
	// @ts-ignore - for some reason ts complains about setBcc
	msg.setBcc(mailData.blindCarbonCopyRecipients.map((rec) => ({ name: rec.name, addr: rec.address })));
	const original = msg.asRaw();
	const blindRecipients = mailData.blindCarbonCopyRecipients.map((recipient) => {
		// @ts-ignore - for some reason ts complains about setBcc
		msg.setBcc({ name: recipient.name, addr: recipient.address });
		return { recipient, content: msg.asRaw() };
	});

	return { original, visibleRecipients, blindRecipients };
};
