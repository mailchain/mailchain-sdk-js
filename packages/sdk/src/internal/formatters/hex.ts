import { decodeHex, decodeHexZeroX } from '@mailchain/encoding';

export function toUint8Array(value: Uint8Array | string | Buffer): Uint8Array {
	if (typeof value === 'string') {
		value = value.trim();
		return value.startsWith('0x') ? decodeHexZeroX(value) : decodeHex(value);
	}

	return value;
}
