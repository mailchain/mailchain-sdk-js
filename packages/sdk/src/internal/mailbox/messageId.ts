import { encodeHex } from '@mailchain/encoding';
import { KeyRing } from '@mailchain/keyring';
import { sha3_256 } from '@noble/hashes/sha3';
import { MailData } from '../formatters/types';

export type MessageIdCreator = (
	params: { mailData: MailData; type: 'sent' } | { mailData: MailData; type: 'received'; owner: string },
) => Promise<string>;

export function createMailchainMessageIdCreator(keyRing: KeyRing): MessageIdCreator {
	return async (params) => {
		let signContent = params.type + params.mailData.id;
		if (params.type === 'received') {
			signContent += params.owner;
		}
		return encodeHex(sha3_256(await keyRing.rootInboxKey().sign(Buffer.from(signContent))));
	};
}
