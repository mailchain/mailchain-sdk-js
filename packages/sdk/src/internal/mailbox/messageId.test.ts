import { AliceED25519PrivateKey, BobED25519PrivateKey } from '@mailchain/crypto/ed25519/test.const';
import { KeyRing } from '@mailchain/keyring';
import { MailData } from '@mailchain/sdk/internal/formatters/types';
import { createMailchainMessageIdCreator } from './messageId';

describe('mailbox message id', () => {
	it('should create unique id for sent messages', async () => {
		const aliceHasher = createMailchainMessageIdCreator(KeyRing.fromPrivateKey(AliceED25519PrivateKey));
		const bobHasher = createMailchainMessageIdCreator(KeyRing.fromPrivateKey(BobED25519PrivateKey));

		const aliceHash = await aliceHasher({
			mailData: { id: '123@mailchain.local' } as MailData,
			type: 'sent',
		});
		const bobHash = await bobHasher({ mailData: { id: '123@mailchain.local' } as MailData, type: 'sent' });

		expect(aliceHash).not.toEqual(bobHash);
		expect(aliceHash).toMatchSnapshot('aliceSentHash');
		expect(bobHash).toMatchSnapshot('bobSentHash');
	});

	it('should create unique id for received messages', async () => {
		const aliceHasher = createMailchainMessageIdCreator(KeyRing.fromPrivateKey(AliceED25519PrivateKey));

		const recipientHash1 = await aliceHasher({
			mailData: { id: '123@mailchain.local' } as MailData,
			type: 'received',
			owner: '0x4321@ethereum.mailchain.local',
		});
		const recipientHash2 = await aliceHasher({
			mailData: { id: '123@mailchain.local' } as MailData,
			type: 'received',
			owner: '0x1234@ethereum.mailchain.local',
		});

		expect(recipientHash1).not.toEqual(recipientHash2);
		expect(recipientHash1).toMatchSnapshot('recipientHash1');
		expect(recipientHash2).toMatchSnapshot('recipientHash2');
	});
});
