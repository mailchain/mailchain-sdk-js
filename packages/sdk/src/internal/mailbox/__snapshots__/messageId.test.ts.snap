// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mailbox message id should create unique id for received messages: recipientHash1 1`] = `"5118b83a98441db6bf202a43c7dc6f002418f2f23dd583e14a71a1a81db5a77a"`;

exports[`mailbox message id should create unique id for received messages: recipientHash2 1`] = `"875bb458b4a5493e081f1499581d79a7203632d2f1197a975a5d99e8c9607c29"`;

exports[`mailbox message id should create unique id for sent messages: aliceSentHash 1`] = `"112574bb1fcf3fe670bdcd14e2dda34de77343509adb8f9a68dc9b16a0414602"`;

exports[`mailbox message id should create unique id for sent messages: bobSentHash 1`] = `"f89aac691e1de8bae733101fd7b73bf38aae055e0d510a6cc6119a8b32a88ba2"`;
